src/foam/box/socket/SocketConnectionBox.js

```diff
@@ -0,0 +1,272 @@
+    'java.nio.charset.StandardCharsets',
+    'java.util.Map',
+    'java.util.HashMap',
+    'java.util.Collections',
+    'java.util.concurrent.atomic.AtomicLong'
+  ],
+    
+  constants: [
+    {
+      name: 'REPLY_BOX_ID',
+      value: 'REPLY_BOX_ID',
+      type: 'String'
+    }
+  ],
+
+  properties: [
+    {
+      documentation: 'managed by SocketConnectionBoxManager',
+      name: 'key',
+      class: 'String'
+    },
+    {
+      name: 'host',
+      class: 'String'
+    },
+    {
+      name: 'port',
+      class: 'Int'
+    },
+    {
+      name: 'socket',
+      class: 'Object',
+      visibility: 'HIDDEN'
+    },
+    {
+      documentation: 'Set to false when send exits, triggering execute to exit',
+      name: 'valid',
+      class: 'Boolean',
+      value: true,
+      visibility: 'HIDDEN'
+    },
+    {
+      name: 'replyBoxes',
+      class: 'Map',
+      javaFactory: `return new HashMap();`,
+      visibility: 'HIDDEN',
+    },
+    {
+      name: 'logger',
+      class: 'FObjectProperty',
+      of: 'foam.nanos.logger.Logger',
+      visibility: 'HIDDEN',
+      javaFactory: `
+        return new PrefixLogger(new Object[] {
+          this.getClass().getSimpleName(),
+          getHost(),
+          getPort()
+        }, (Logger) getX().get("logger"));
+      `
+    }
+  ],
+
+  axioms: [
+    {
+      name: 'javaExtras',
+      buildJavaClass: function(cls) {
+        cls.extras.push(foam.java.Code.create({
+          data: `
+  public SocketConnectionBox(X x, String key, Socket socket, String host, int port)
+    throws IOException
+  {
+    setX(x);
+    setKey(key);
+    setHost(host);
+    setPort(port);
+    setSocket(socket);
+
+    out_ = new DataOutputStream(new BufferedOutputStream(socket.getOutputStream()));
+    in_ = new DataInputStream(socket.getInputStream());
+  }
+
+  protected DataInputStream in_;
+  protected DataOutputStream out_;
+  private static AtomicLong replyBoxId_ = new AtomicLong(0);
+
+  protected static final ThreadLocal<foam.lib.formatter.FObjectFormatter> formatter_ = new ThreadLocal<foam.lib.formatter.FObjectFormatter>() {
+    @Override
+    protected foam.lib.formatter.JSONFObjectFormatter initialValue() {
+      foam.lib.formatter.JSONFObjectFormatter formatter = new foam.lib.formatter.JSONFObjectFormatter();
+      formatter.setQuoteKeys(true);
+      formatter.setPropertyPredicate(new foam.lib.ClusterPropertyPredicate());
+      return formatter;
+    }
+
+    @Override
+    public foam.lib.formatter.FObjectFormatter get() {
+      foam.lib.formatter.FObjectFormatter formatter = super.get();
+      formatter.reset();
+      return formatter;
+    }
+  };
+        `
+        }));
+      }
+    }
+  ],
+
+  methods: [
+    {
+      name: 'send',
+      javaCode: `
+      PM pm = PM.create(getX(), this.getClass().getSimpleName(), getHost()+":"+getPort()+":send");
+      Box replyBox = (Box) msg.getAttributes().get("replyBox");
+      if ( replyBox != null ) {
```

Comment:
why does it seem like the code below should only be executed if replyBox == null, not replyBox != null? what is happening when the replyBox is null?